### PR TITLE
Feature "Commit Date for Branch"

### PR DESCRIPTION
**Pull request by pfeuffer**
## Proposed changes

Show the date of the last commit for branches and split the branches overview into "active" and "stale" branches depending on their age releative to the default branch.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] PR is well described
- [x] Related issues linked to PR if existing and labels set
- [x] Target branch is not master (in most cases develop should bet the target of choice) 
- [x] Code does not conflict with target branch
- [x] New code is covered with unit tests
- [x] CHANGELOG.md updated
- [x] Definition of Done's fulfilled: [DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/definition-of-done.md) // [UI DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/ui-dod.md)
- [x] Documentation updated (only necessary for new features or changed behaviour)

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: docs/de/user/repo/branches.md [hunk 1]
```diff
@@ -3,9 +3,11 @@ title: Repository
 subtitle: Branches
 ---
 ### Übersicht
-Auf der Branches-Übersicht sind die bereits existierenden Branches aufgeführt. Bei einem Klick auf einen Branch wird man zur Detailseite des Branches weitergeleitet. 
+Auf der Branches-Übersicht sind die bereits existierenden Branches aufgeführt. Bei einem Klick auf einen Branch wird man zur Detailseite des Branches weitergeleitet.
+Die Branches sind in zwei Listen aufgeteilt: Unter "Aktive Branches" sind Branches aufgelistet, deren letzter Commit
+nicht 30 Tage älter als derjenige des Default-Branches ist. Alle weiteren Branches sind in der Liste "Stale Branches" zu finden.
```
Comment (eheimbuch) [inline docs/de/user/repo/branches.md:8]:
```suggestion
Die Branches sind in zwei Listen aufgeteilt: Unter "Aktive Branches" sind Branches aufgelistet, deren letzter Commit
nicht 30 Tage älter als der Stand des Default-Branches ist. Alle älteren Branches sind in der Liste "Stale Branches" zu finden.
```

File: docs/en/user/repo/branches.md [hunk 2]
```diff
@@ -4,6 +4,8 @@ subtitle: Branches
 ---
 ### Overview
 The branches overview shows the branches that are already existing. By clicking on a branch, the details page of the branch is shown.
+Branches are split into two lists: Branches whose last commits are at most 30 days older than the one of the default
+branch are listed in "Active Branches". The other ones can be found in "Stale Branches".
```
Comment (eheimbuch) [inline docs/en/user/repo/branches.md:8]:
```suggestion
Branches are split into two lists: Branches whose last commits are at most 30 days older than the head of the default
branch are listed in "Active Branches". The older ones can be found in "Stale Branches".
```

File: scm-core/src/main/java/sonia/scm/repository/api/BranchXDaysOlderThanDefaultStaleComputer.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  public BranchXDaysOlderThanDefaultStaleComputer(int amountOfDays) {
+    this.amountOfDays = amountOfDays;
+  }
+
+  @Override
+  @SuppressWarnings("java:S3655") // we check "isPresent" for both dates, but due to the third check sonar does not get it
+  public boolean computeStale(Branch branch, StaleContext context) {
+    Branch defaultBranch = context.getDefaultBranch();
+    if (!branch.isDefaultBranch() && branch.getLastCommitDate().isPresent() && defaultBranch.getLastCommitDate().isPresent()) {
```
Comment (eheimbuch) [inline scm-core/src/main/java/sonia/scm/repository/api/BranchXDaysOlderThanDefaultStaleComputer.java:53]:
```suggestion
    if (shouldCompute(branch, defaultBranch) {
```

Comment (eheimbuch) [reply scm-core/src/main/java/sonia/scm/repository/api/BranchXDaysOlderThanDefaultStaleComputer.java:53]: Wrap your check to increase readability.

File: scm-core/src/main/java/sonia/scm/repository/spi/BranchesCommand.java [hunk 4]
```diff
@@ -41,13 +38,19 @@
...
+      BranchStaleComputer.StaleContext staleContext = new BranchStaleComputer.StaleContext();
+      staleContext.setDefaultBranch(branch);
+      return staleContext;
+    };
+    List<Branch> branches = getBranches();
+    branches.stream()
+      .filter(Branch::isDefaultBranch)
+      .findFirst()
+      .ifPresent(defaultBranch -> branches.forEach(branch -> branch.setStale(computer.computeStale(branch, createContext.apply(defaultBranch)))));
+    return branches;
```
Comment (eheimbuch) [inline scm-core/src/main/java/sonia/scm/repository/spi/BranchesCommand.java:54]: In general this is fine but i find it kinda hard to read. I would have filtered for the default branch, initialized the StaleContext with the default branch and afterwards compute the branches within a foreach loop.

But you can leave it that way.

**Review by eheimbuch — CHANGES_REQUESTED** [review]

**Comment by eheimbuch** [comment]
Please mind the code smells which sonarcloud found.

**Comment by eheimbuch** [comment]
I have not found changes about the branch caching. Does our current caching already work with your changes? Especially the cache invalidation?

**Comment by pfeuffer** [comment]
> I have not found changes about the branch caching. Does our current caching already work with your changes? Especially the cache invalidation?

The cache is invalidated on every push in the `ClearCacheHook`: https://github.com/scm-manager/scm-manager/blob/develop/scm-core/src/main/java/sonia/scm/repository/api/RepositoryServiceFactory.java#L303

**Comment by pfeuffer** [comment]
Tried to improve the readability in `BranchesComputer`. Please have a look at it, @eheimbuch

**Pull request merged**